### ISSUE TITLE
MFA and API Keys in CI/CD

### ISSUE THREAD

**Issue by iMacTia** [rubygems/rubygems#3092]
I'm having issues with Rubygems MFA.
I've previously generated an API Key that I've been using in my CI/CD pipelines to publish new versions of gems.

However, after recently enabling 2FA/MFA, the pipeline has started failing in Travis with the following error:
`Access Denied. Please sign up for an account at https://rubygems.org (Gems::GemError)`

Although I completely agree with 2FA and MFA, I feel like the current implementation doesn't take into consideration automation and CI/CD Pipelines.
In my experience with other services, 2FA and MFA usually do not apply to API Keys. i.e. when using an API Key to authenticate (which is supposed to be long and hard to hack), you can bypass the 2FA/MFA check.

The current options, from what I read in the doc, are the following, but both of them are not good in the case of CI/CD Pipelines:

1. Disable 2FA/MFA checks for API and leave it on for UI only: this is of course insecure and nullifies the extra security as it seems like the CLI uses the API and you can authenticate to the CLI using less secure means other than API Keys (see below for more information).
2. Use the command line `--totp` argument: this works with CLI, but unfortunately is not possible to the top of my knowledge in CI/CD Pipelines, as I have no idea on how to make them sync with the Authenticator App on my phone.

The confusion on point 1 comes with the CLI.
On my laptop, the first time I used the cli I remember my username and password were asked:

> After creating the account, use your email and password when pushing the gem. (RubyGems saves the credentials in ~/.gem/credentials for you so you only need to log in once.)

Now, that is of course unsafe, so I'd ideally like the CLI to request the MFA code in order to do anything. However, according to the documentation, that is only possible by enabling MFA for both UI and API (https://guides.rubygems.org/using-mfa-in-command-line/).

My suggestion would be to provide an extra option so that 2FA/MFA can be enabled for BOTH API and UI, but ignored when API authentication is done through API Keys only.

Any help would be greatly appreciated!

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [x] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

**Comment by mislav** [comment]
The way I understand it: "UI" = Rubygems.org web; "API" = `gem push` + other programmatic usage.

Even though `gem push` offers to authenticate using email + password, it converts that to an API key which it caches; see `~/.gem/credentials`.

Personally, I do like to have both UI & API protected by MFA because I wouldn't like that an attacker who gains my password be able push new versions of my gems using `gem push`.

However, the downside is that I can't publish new gems from automation (CI/CD) anymore. I'm not sure what the correct soluton would be. Perhaps there could be a special API key that's exempt from MFA? I don't know if offering that would defeat the whole point of MFA.

**Comment by iMacTia** [comment]
Gentle bump on this one. Still an issue for me. Is there a working solution for this now?

**Comment by simi** [comment]
> In my experience with other services, 2FA and MFA usually do not apply to API Keys. i.e. when using an API Key to authenticate (which is supposed to be long and hard to hack), you can bypass the 2FA/MFA check.

Can you point me to some docs another service provide for this? We can take a look and inspire in there. Also what do you suggest, to implement at rubygems.org side?

> Is there a working solution for this now?

Would it be possible to create special CI user at rubygems.org having MFA disabled, add it as an owner to your gem, ensure that account is extra secured on your side (even without MFA) and use that on CI/CD for pushes? That way your personal account could stay with MFA, but you should be able to push on CI/CD on alternative account.

_PS: I think we should move this issue to rubygems.org repo._

**Comment by iMacTia** [comment]
@simi @deivid-rodriguez the standard in this situation is to allow users to generate "application tokens". These are longer passwords supposed to be used with specific applications.
i.e. if I authenticate via the CLI I can provide my username and a generated token in place of the password. When that happens, the MFA code is not requested. If you instead provide the standard password, the MFA code will be necessary.

Top of my mind, both GitHub and Google implement this, although I'm sure I've seen this same practice quite a lot.
GitHub calls them "Personal Access Tokens" and has already deprecated standard password authentication: https://docs.github.com/en/rest/overview/other-authentication-methods#working-with-two-factor-authentication

**Comment by simi** [comment]
So for now it should be possible to temporarily simulate "application token" with new user with disaled MFA just for this purpose, right?

**Comment by iMacTia** [comment]
Yes correct, I can potentially create a new user with a really long password and MFA disabled, and use that one to push gems. I won't do that because my password is already reasonably long so I'm temporarily OK with MFA disabled until this is solved.
Managing multiple users is definitely more painful and not a long term solution, but at least you could suggest this solution for the time being (or increase password size).

BTW this kind of solution might work for Rubygems but it's worth nothing that it would not normally work for others (e.g. GitHub) when there are organisational constraints. For example, on GitHub, I could enforce everyone in my organisation to have MFA enabled, and this would have to apply to all users, including the dummy one you suggest to create.

If Rubygems were to introduce a security feature like "only allow MFA-protected owners to publish", then your solution wouldn't work anymore.

**Comment by deivid-rodriguez** [comment]
@iMacTia Yeah, I understand this issue and I agree it would be useful to have something like github's "Personal Access Token"'s in rubygems.org to make it possible to get both MFA and `gem push` automation at the same time.

However, I don't think exempting current API keys from MFA is a good solution since API keys currently give you access to essentially every API operation (API keys were recently scoped by gem operation, but even with that, an API key with the "gem push" scope still gives you access to pushing any gem owned by the key owner). Doing that would be essentially be removing the feature of protecting API keys with MFA.

I think we should only allow a key to be exempted from MFA if it's scoped by both gem and gem operation. So if you generate a "personal access token" that only allows to `gem push` new versions of gem `foo`, then it would be fine that this kind of key is excluded from MFA.

I'll transfer this ticket to rubygems.org because I also think it belongs there, and also to get feedback from rubygems.org folks.
